[GSB] When adding same-type requirements pick representative based on…

xedin opened the pull request:
… canonical order

Instead of trying to order based on the "nested depth", let's
always prefer canonical ordering of type parameters when it comes
to picking representative equivalence class.

Resolves: rdar://problem/45957015

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[GSB] When adding same-type requirements pick representative based on canonical order

Instead of trying to order based on the "nested depth", let's
always prefer canonical ordering of type parameters when it comes
to picking representative equivalence class.

Resolves: rdar://problem/45957015

xedin commented on test/Generics/same_type_constraints.swift:
```diff
@@ -351,11 +351,11 @@ func intercomponentSameComponents<T: P10>(_: T)
 
 func intercomponentMoreThanSpanningTree<T: P10>(_: T)
-  where T.A == T.B,
+  where T.A == T.B, // expected-note {{previous same-type constraint 'T.A' == 'T.B' written here}}
```
This is definitely not ideal because these requirements should be diagnosed in source order, but I think I can fix that...

xedin commented:
@swift-ci please test source compatibility

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -4889,15 +4889,12 @@ GenericSignatureBuilder::addSameTypeRequirementBetweenTypeParameters(
-  unsigned nestingDepth2 = T2->getNestingDepth();
-  if (nestingDepth2 < nestingDepth1) {
+  // Pick representative based on the canonical ordering of the type parameters.
+  if (compareDependentTypes(depType2, depType1) < 0) {
```
This is a very nice simplification, thank you!

DougGregor commented on test/decl/protocol/req/associated_type_ambiguity.swift:
```diff
@@ -48,7 +48,7 @@ extension P2 where Self : P3, T == Int {
 
 extension P2 where Self : P3 {
-  func takeT1(_: T) {}
+  func takeT1(_: T) {} // expected-error {{'T' is ambiguous for type lookup in this context}}
```
I feel like we've tripped over this specific bug in a number of places, where unqualified lookup is ambiguous but shouldn't be. I suspect your change isn't making it much worse in practice.

xedin replied:
Sounds good! I'm going to work on other diagnostic...

xedin committed:
[GSB/Diagnostics] Improve redundant same-type constraint diagnostics

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -6905,34 +6907,121 @@ void GenericSignatureBuilder::checkSameTypeConstraints(
+    for (unsigned i : indices(intercomponentEdges))
+      sourceOrderedEdges.push_back(i);
+
+    std::sort(sourceOrderedEdges.begin(), sourceOrderedEdges.end(),
```
Is there any way we can avoid sorting in the common case where there aren't any diagnostics to emit? The `GenericSignatureBuilder` is very much a hot path in the type checker.

xedin replied:
Sure, I can use `array_pod_sort` there, but it seems like this is not a common path because it's been determined that we have more inter-component edges than components and it has to be diagnosed?...

DougGregor replied:
Hmm... I see what you mean, okay!

xedin replied:
Moved to `llvm::array_pod_sort`, thanks for a suggestion!

xedin committed:
[GSB/Diagnostics] Use `llvm::array_pod_sort` to sort "as written" edges

Instead of using `std::sort`, transform source-ordered edge vector
to reference `IntercomponentEdge *` instead of their indices in
`intercomponentEdges` vector and use `llvm::array_pod_sort` to sort them.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

DougGregor commented:
How interesting! This pull request broke the tests I just tried to add for https://bugs.swift.org/browse/SR-8767, which had been fixed on master before that (hence my attempt at merging https://github.com/apple/swift/pull/20795). I wonder if we were getting lucky, or if there's something deeper that we didn't notice.
